### DISCUSSION TITLE
为什么我的pdf目录标签前面都出现一个“-4”啊

### DISCUSSION THREAD

**Discussion by bxdd** [nju-lug/NJUThesis#237] [Q&A] [answered]
如图，求可能的解决方案。QWQ
![image](https://github.com/nju-lug/NJUThesis/assets/45119470/80296236-0ac5-4040-8d89-e09fb292ea62)

**Comment by bxdd** [comment]
经过探索，发现是这句命令的原因。有什么其他方法改变目录里的字号吗？
![image](https://github.com/nju-lug/NJUThesis/assets/45119470/1cda95f4-1ff1-4e1a-aa2f-bfd5949b8c83)

**Comment by atxy-blip** [comment]
解答见 #236。因为这个格式命令是写进 `.aux` 的，不允许加参数。模板也是定义了一个 `\bigger` 套了一层。

**Reply by bxdd** [reply]
那请问有什么方法可以修改目录里面chapter的字号吗？

**Reply by bxdd** [reply]
因为我修改了正文里面chapter的字号，然后就发现目录里面的chapter字号也变了。所以我就想把目录里面的chapter字号缩小回去。

**Reply by atxy-blip** [reply]
> 模板也是定义了一个 `\bigger` 套了一层。

详细来说就是
```tex
\newcommand\foo{\zihao{-4}}
\njusetformat{chapterintoc}{\foo\heiti}
```

**Reply by bxdd** [reply]
感谢了qwq

**Reply by bxdd** [reply]
还是不行啊，我都改成这样了，编译缓存也关了，还是一样有-4：
![image](https://github.com/nju-lug/NJUThesis/assets/45119470/edd315f7-9d65-4049-82be-c7b8b59620a6)

**Reply by atxy-blip** [reply, marked as answer]
忘了 `\newcommand` 不带保护，
```tex
\NewDocumentCommand{\chapterzihao}{}{\zihao{1}}
\NewDocumentCommand{\sectionzihao}{}{\zihao{-2}}
\NewDocumentCommand{\subsubzihao}{}{\zihao{4}}
\NewDocumentCommand{\parazihao}{}{\zihao{-4}}
\njusetformat{chapter}{\protect\chapterzihao\heiti\centering}
\njusetformat{section}{\sectionzihao\heiti}
\njusetformat{paragraph}{\parazihao\heiti}
\njusetformat{subsubsection}{\subsubzihao\heiti}
\njusetformat{chapterintoc}{\parazihao\heiti}
\njusetformat{toctitle}{\protect\chapterzihao\heiti\centering}
```
顺便建议用 markdown 代码块来呈现代码，而不是截图（

**Reply by bxdd** [reply]
谢谢！

**Comment by muzimuzhi** [comment]
使用 `\njusetformat{chapterintoc}{\zihao{-4}\heiti}`，导致 PDF 目录标签里，chapter 对应的标签前面都出现一个 "-4"，是因为
1. `njuthesis` 现在把 `chapterintoc` 等样式存在 token list variable 里，然后把它们直接插入到章节标题的开头，导致样式也是标题的一部分。最好不要这样。
1. 于是，包含样式的章节标题中会有 `hyperref` 不认识的命令。这些命令在生成 PDF 目录时会自动忽略，但它们的参数会留下。所以 `\zihao{-4}` 会生成 warning
   ```tex
   Package hyperref Warning: Token not allowed in a PDF string (Unicode):
   (hyperref)                removing `\zihao' on input line 21.
   ```
   同时把参数 `-4` 留在 PDF 目录中。

https://github.com/nju-lug/NJUThesis/discussions/237#discussioncomment-8655348 里的做法，是让样式由不可展开的、无参数的命令组成。这样能避免把参数留在 PDF 目录中，但 warning 依然生成。

最好的办法，是不要把样式插入到章节标题内容里去，把两者分离。

次好的办法，可以让 `hyperref` 认识要插入的命令，如
```tex
\ExplSyntaxOn
\pdfstringdefDisableCommands
  {
    \tl_clear:N \c__nju_fmt_chapterintoc_tl
  }
\ExplSyntaxOff
```
备注：
- 从功能来看，更适合使用 `\cs_set_eq:NN` 到 `\prg_do_nothing:` 或 `\relax` 之类。
- 这个 `\c_..._tl` 会被修改，所以应命名为 `\l_..._tl`，见 https://github.com/nju-lug/NJUThesis/issues/240。

------

顺便，https://github.com/nju-lug/NJUThesis/discussions/237#discussioncomment-8655348 例子里最后一行的 `\protect` 可以删去。

> ```tex
> \njusetformat{toctitle}{\protect\chapterzihao\heiti\centering}
> ```

**Comment by atxy-blip** [comment]
> `njuthesis` 现在把 `chapterintoc` 等样式存在 token list variable 里，然后把它们直接插入到章节标题的开头，导致样式也是标题的一部分。最好不要这样。

当时翻了一遍 `ctex` 手册，发现没有办法不引入额外宏包实现目录条目样式的单独调整，最接近的选项是 `\ctexset{chapter/tocline=...}`，就简单 hack 了一下：

<https://github.com/nju-lug/NJUThesis/blob/8e3eaa1b79140761fce0f7a52a6d9009a886c156/source/njuthesis.dtx#L5554-L5554>

我先采用下策，`\cs_set_eq:NN \zihao \use_none:n`。至于上策，我对目录这块的实现不算很熟悉，还不知道怎么能做这个样式分离 orz

**Reply by muzimuzhi** [reply]
https://github.com/CTeX-org/ctex-kit/issues/588 里有一点关于目录宏包的介绍
